### This Workbook is structured as follows:

### 1) Importing libraries and data
### 2) Cleaning and aggregating the data for analysis
### 3) Merging the data 

# 1) Importing Libraries and Data

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Setting path to data
path = r'C:\Users\mmoss\07-02-2022 NHL Data Analysis'


In [3]:
# Importing game.csv
df_game = pd.read_csv(os.path.join(path, '02 Data', 'Original', 'game.csv'), index_col = False)

In [4]:
# Showing the head
df_game.head()

game_id    season type         date_time_GMT  away_team_id  \
0  2016020045  20162017    R  2016-10-19T00:30:00Z             4   
1  2017020812  20172018    R  2018-02-07T00:00:00Z            24   
2  2015020314  20152016    R  2015-11-24T01:00:00Z            21   
3  2015020849  20152016    R  2016-02-17T00:00:00Z            52   
4  2017020586  20172018    R  2017-12-30T03:00:00Z            20   

   home_team_id  away_goals  home_goals       outcome home_rink_side_start  \
0            16           4           7  home win REG                right   
1             7           4           3   away win OT                 left   
2            52           4           1  away win REG                right   
3            12           1           2  home win REG                right   
4            24           1           2  home win REG                 left   

            venue           venue_link   venue_time_zone_id  \
0   United Center  /api/v1/venues/null      America/Chicago   
1  KeyBank Center  /api/v1/venues/null     America/New_York   
2      MTS Centre  /api/v1/venues/null     America/Winnipeg   
3       PNC Arena  /api/v1/venues/null     America/New_York   
4    Honda Center  /api/v1/venues/null  America/Los_Angeles   

   venue_time_zone_offset venue_time_zone_tz  
0                      -5                CDT  
1                      -4                EDT  
2                      -5                CDT  
3                      -4                EDT  
4                      -7                PDT

In [5]:
# Importing game_skater_stats.csv
df_skater_stats = pd.read_csv(os.path.join(path, '02 Data', 'Original', 'game_skater_stats.csv'), index_col = False)

In [6]:
# Showing the head
df_skater_stats.head()

game_id  player_id  team_id  timeOnIce  assists  goals  shots  hits  \
0  2016020045    8468513        4        955        1      0      0   2.0   
1  2016020045    8476906        4       1396        1      0      4   2.0   
2  2016020045    8474668        4        915        0      0      1   1.0   
3  2016020045    8473512        4       1367        3      0      0   0.0   
4  2016020045    8471762        4        676        0      0      3   2.0   

   powerPlayGoals  powerPlayAssists  ...  faceoffTaken  takeaways  giveaways  \
0               0                 0  ...             0        1.0        1.0   
1               0                 0  ...             0        1.0        2.0   
2               0                 0  ...             0        2.0        0.0   
3               0                 2  ...            27        0.0        0.0   
4               0                 0  ...             0        0.0        1.0   

   shortHandedGoals  shortHandedAssists  blocked  plusMinus  evenTimeOnIce  \
0                 0                   0      1.0          1            858   
1                 0                   0      2.0          0           1177   
2                 0                   0      0.0         -1            805   
3                 0                   0      0.0         -1           1083   
4                 0                   0      0.0         -1            613   

   shortHandedTimeOnIce  powerPlayTimeOnIce  
0                    97                   0  
1                     0                 219  
2                     0                 110  
3                    19                 265  
4                    63                   0  

[5 rows x 22 columns]

In [7]:
# Importing game_teams_wins.csv
df_team_stats = pd.read_csv(os.path.join(path, '02 Data', 'Original', 'game_teams_wins.csv'), index_col = False)

In [8]:
# Showing the head
df_team_stats.head()

team_id  number_of_wins
0        1            1691
1        2            1743
2        3            1768
3        4            1735
4        5            1786

# 2) Cleaning and aggregating the data

### Checking with describe

In [9]:
# Checking df_game data frame using df.describe
df_game.describe()

game_id        season  away_team_id  home_team_id    away_goals  \
count  2.630500e+04  2.630500e+04  26305.000000  26305.000000  26305.000000   
mean   2.010765e+09  2.010945e+07     16.879377     16.881429      2.687968   
std    6.073568e+06  6.073849e+04     11.184920     11.205625      1.619068   
min    2.000020e+09  2.000200e+07      1.000000      1.000000      0.000000   
25%    2.006020e+09  2.006201e+07      8.000000      8.000000      1.000000   
50%    2.011021e+09  2.011201e+07     16.000000     16.000000      3.000000   
75%    2.016030e+09  2.016202e+07     24.000000     24.000000      4.000000   
max    2.019041e+09  2.019202e+07     90.000000     90.000000     11.000000   

         home_goals  venue_time_zone_offset  
count  26305.000000            26305.000000  
mean       2.958981               -5.066147  
std        1.689031                1.210145  
min        0.000000               -8.000000  
25%        2.000000               -6.000000  
50%        3.000000               -5.000000  
75%        4.000000               -4.000000  
max       12.000000               -4.000000

Everything seems normal here.

In [10]:
# Checking df_skater_stats
df_skater_stats.describe()

game_id     player_id        team_id      timeOnIce  \
count  9.458300e+05  9.458300e+05  945830.000000  945830.000000   
mean   2.010763e+09  8.469436e+06      16.862985     988.705195   
std    6.072766e+06  7.299758e+03      11.146468     310.968101   
min    2.000020e+09  8.444894e+06       1.000000       0.000000   
25%    2.006020e+09  8.466181e+06       8.000000     787.000000   
50%    2.011021e+09  8.470626e+06      16.000000     999.000000   
75%    2.016030e+09  8.475169e+06      24.000000    1198.000000   
max    2.019041e+09  8.481813e+06      90.000000    4316.000000   

             assists          goals          shots           hits  \
count  945830.000000  945830.000000  945830.000000  547723.000000   
mean        0.262792       0.154644       1.664201       1.304696   
std         0.533979       0.408620       1.547529       1.462937   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       1.000000       1.000000   
75%         0.000000       0.000000       2.000000       2.000000   
max         5.000000       5.000000      16.000000      15.000000   

       powerPlayGoals  powerPlayAssists  ...   faceoffTaken      takeaways  \
count   945830.000000     945830.000000  ...  945830.000000  547723.000000   
mean         0.037124          0.070543  ...       3.345550       0.391313   
std          0.196857          0.276856  ...       6.176884       0.671742   
min          0.000000          0.000000  ...       0.000000       0.000000   
25%          0.000000          0.000000  ...       0.000000       0.000000   
50%          0.000000          0.000000  ...       0.000000       0.000000   
75%          0.000000          0.000000  ...       3.000000       1.000000   
max          4.000000          5.000000  ...      46.000000       8.000000   

           giveaways  shortHandedGoals  shortHandedAssists        blocked  \
count  547723.000000     945830.000000       945830.000000  547723.000000   
mean        0.475580          0.004924            0.005195       0.798327   
std         0.765434          0.070568            0.072288       1.095038   
min         0.000000          0.000000            0.000000       0.000000   
25%         0.000000          0.000000            0.000000       0.000000   
50%         0.000000          0.000000            0.000000       0.000000   
75%         1.000000          0.000000            0.000000       1.000000   
max         8.000000          2.000000            2.000000      15.000000   

           plusMinus  evenTimeOnIce  shortHandedTimeOnIce  powerPlayTimeOnIce  
count  945830.000000  945830.000000         945830.000000       945830.000000  
mean       -0.007727     807.715360             80.117924          100.871911  
std         1.094068     245.388009             95.408978          118.131458  
min        -7.000000       0.000000              0.000000            0.000000  
25%        -1.000000     650.000000              0.000000            0.000000  
50%         0.000000     809.000000             47.000000           60.000000  
75%         1.000000     968.000000            133.000000          169.000000  
max         7.000000    3411.000000            908.000000         1205.000000  

[8 rows x 22 columns]

Everything seems fine here. 

### Checking for Mixed Type Columns

In [11]:
# Checking df_game for mixed type columns
for col in df_game.columns.tolist():
  weird = (df_game[[col]].applymap(type) != df_game[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_game[weird]) > 0:
    print (col)

home_rink_side_start


home_rink_side_start is a mixed type column

In [12]:
# Changing the columns values to a string
df_game['home_rink_side_start'] = df_game['home_rink_side_start'].astype('str')

In [13]:
# Checking mixed type columns again
for col in df_game.columns.tolist():
  weird = (df_game[[col]].applymap(type) != df_game[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_game[weird]) > 0:
    print (col)

No column name was printed. Success!

In [14]:
# Checking df_skater_stats for mixed type columns
for col in df_skater_stats.columns.tolist():
  weird = (df_skater_stats[[col]].applymap(type) != df_skater_stats[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_skater_stats[weird]) > 0:
    print (col)

No column name was printed. Success!

### Checking for missing values

In [15]:
# Checking df_game
df_game.isnull().sum()

game_id                   0
season                    0
type                      0
date_time_GMT             0
away_team_id              0
home_team_id              0
away_goals                0
home_goals                0
outcome                   0
home_rink_side_start      0
venue                     0
venue_link                0
venue_time_zone_id        0
venue_time_zone_offset    0
venue_time_zone_tz        0
dtype: int64

No missing values in df_game

In [16]:
# Checking df_skater_stats
df_skater_stats.isnull().sum()

game_id                      0
player_id                    0
team_id                      0
timeOnIce                    0
assists                      0
goals                        0
shots                        0
hits                    398107
powerPlayGoals               0
powerPlayAssists             0
penaltyMinutes               0
faceOffWins                  0
faceoffTaken                 0
takeaways               398107
giveaways               398107
shortHandedGoals             0
shortHandedAssists           0
blocked                 398107
plusMinus                    0
evenTimeOnIce                0
shortHandedTimeOnIce         0
powerPlayTimeOnIce           0
dtype: int64

df_skater_stats has 398,107 missing values in the hits, takeaways, giveaways and blocked columns. This needs to be fixed.

Since the hits column does not have many values exceeding 10, and most hang around 1-3 I think using the mean of 1.3 would be appropriate here for imputing.

In [17]:
#Replacing the hits column with the average
df_skater_stats['hits'] = df_skater_stats['hits'].fillna(df_skater_stats['hits'].mean())

Since the takeaways column does not usually exceed 4 I am going to impute using the average

In [18]:
#Replacing the takeaways column with the average
df_skater_stats['takeaways'] = df_skater_stats['takeaways'].fillna(df_skater_stats['takeaways'].mean())

Since there are a lot of 0s in the give away column I think the average of 0.4 is appropriate here.

In [19]:
#Replacing the giveaways column with the average
df_skater_stats['giveaways'] = df_skater_stats['giveaways'].fillna(df_skater_stats['giveaways'].mean())

The average seems appropriate for the blocked column as well. Going to impute the average.

In [20]:
#Replacing the blocked column with the average
df_skater_stats['blocked'] = df_skater_stats['blocked'].fillna(df_skater_stats['blocked'].mean())

In [21]:
# Re-Checking df_skater_stats
df_skater_stats.isnull().sum()

game_id                 0
player_id               0
team_id                 0
timeOnIce               0
assists                 0
goals                   0
shots                   0
hits                    0
powerPlayGoals          0
powerPlayAssists        0
penaltyMinutes          0
faceOffWins             0
faceoffTaken            0
takeaways               0
giveaways               0
shortHandedGoals        0
shortHandedAssists      0
blocked                 0
plusMinus               0
evenTimeOnIce           0
shortHandedTimeOnIce    0
powerPlayTimeOnIce      0
dtype: int64

Success!

### Checking for Duplicates

In [22]:
# Checking the skaters df
df_dups_skaters = df_skater_stats[df_skater_stats.duplicated()]

In [23]:
pd.set_option('display.max_columns', None)

In [24]:
#running it
df_dups_skaters

game_id  player_id  team_id  timeOnIce  assists  goals  shots  \
760942  2019020001    8480073        9        957        0      0      1   
760943  2019020001    8476919        9       1061        1      0      1   
760944  2019020001    8477407        9       1013        0      0      3   
760945  2019020001    8476419        9        792        1      0      0   
760946  2019020001    8468493        9       1053        0      0      0   
...            ...        ...      ...        ...      ...    ...    ...   
945825  2018030417    8475807        6        501        0      0      1   
945826  2018030417    8475149        6       1062        0      0      2   
945827  2018030417    8475745        6        913        0      0      2   
945828  2018030417    8478443        6       1006        0      0      1   
945829  2018030417    8478046        6        909        0      0      2   

        hits  powerPlayGoals  powerPlayAssists  penaltyMinutes  faceOffWins  \
760942   4.0               0                 0               0            0   
760943   5.0               0                 0               0            9   
760944   4.0               0                 0               0            1   
760945   1.0               0                 0               0            8   
760946   2.0               0                 0               0            0   
...      ...             ...               ...             ...          ...   
945825   2.0               0                 0               0            0   
945826   0.0               0                 0               0            1   
945827   3.0               0                 0               0            4   
945828   1.0               0                 0               0            0   
945829   0.0               0                 0               0            0   

        faceoffTaken  takeaways  giveaways  shortHandedGoals  \
760942             0        0.0        1.0                 0   
760943            15        0.0        1.0                 0   
760944             1        0.0        0.0                 0   
760945            15        1.0        0.0                 0   
760946             0        0.0        1.0                 0   
...              ...        ...        ...               ...   
945825             0        0.0        0.0                 0   
945826             1        1.0        0.0                 0   
945827            11        1.0        0.0                 0   
945828             0        0.0        1.0                 0   
945829             0        1.0        1.0                 0   

        shortHandedAssists  blocked  plusMinus  evenTimeOnIce  \
760942                   0      2.0          0            820   
760943                   0      2.0          1            723   
760944                   0      0.0          1            610   
760945                   0      0.0          1            576   
760946                   0      2.0          0            770   
...                    ...      ...        ...            ...   
945825                   0      0.0         -1            501   
945826                   0      0.0         -1            957   
945827                   0      0.0          0            913   
945828                   0      2.0         -1           1006   
945829                   0      0.0          0            909   

        shortHandedTimeOnIce  powerPlayTimeOnIce  
760942                     0                 137  
760943                   149                 189  
760944                   200                 203  
760945                   215                   1  
760946                   283                   0  
...                      ...                 ...  
945825                     0                   0  
945826                     0                 105  
945827                     0                   0  
945828                     0                   0  
945829                     0          

Duplicates are necessary in this scenario. Leaving them in. 

In [25]:
#Checking the game df
df_dups_game = df_game[df_game.duplicated()]

In [26]:
#running it
df_dups_game

game_id    season type         date_time_GMT  away_team_id  \
21164  2019020001  20192020    R  2019-10-02T23:00:00Z             9   
21165  2019020002  20192020    R  2019-10-03T00:00:00Z            15   
21166  2019020003  20192020    R  2019-10-03T02:00:00Z            23   
21167  2019020004  20192020    R  2019-10-03T02:30:00Z            28   
21176  2019020005  20192020    R  2019-10-03T23:00:00Z            13   
...           ...       ...  ...                   ...           ...   
26296  2018030413  20182019    P  2019-06-02T00:00:00Z             6   
26298  2018030414  20182019    P  2019-06-04T00:00:00Z             6   
26300  2018030415  20182019    P  2019-06-07T00:00:00Z            19   
26302  2018030416  20182019    P  2019-06-10T00:00:00Z             6   
26304  2018030417  20182019    P  2019-06-13T00:00:00Z            19   

       home_team_id  away_goals  home_goals       outcome  \
21164            10           3           5  home win REG   
21165            19           3           2   away win OT   
21166            22           2           3  home win REG   
21167            54           1           4  home win REG   
21176            14           2           5  home win REG   
...             ...         ...         ...           ...   
26296            19           7           2  away win REG   
26298            19           2           4  home win REG   
26300             6           2           1  away win REG   
26302            19           5           1  away win REG   
26304             6           4           1  away win REG   

      home_rink_side_start              venue           venue_link  \
21164                right   Scotiabank Arena  /api/v1/venues/null   
21165                 left  Enterprise Center  /api/v1/venues/5076   
21166                 left       Rogers Place  /api/v1/venues/5100   
21167                  nan     T-Mobile Arena  /api/v1/venues/5178   
21176                 left       Amalie Arena  /api/v1/venues/5017   
...                    ...                ...                  ...   
26296                 left  Enterprise Center  /api/v1/venues/5076   
26298                 left  Enterprise Center  /api/v1/venues/5076   
26300                 left          TD Garden  /api/v1/venues/5085   
26302                 left  Enterprise Center  /api/v1/venues/5076   
26304                 left          TD Garden  /api/v1/venues/5085   

        venue_time_zone_id  venue_time_zone_offset venue_time_zone_tz  
21164      America/Toronto                      -5                EST  
21165      America/Chicago                      -6                CST  
21166     America/Edmonton                      -7                MST  
21167  America/Los_Angeles                      -8                PST  
21176     America/New_York                      -5                EST  
...                    ...                     ...                ...  
26296      America/Chicago                      -5                CDT  
26298      America/Chicago                      -6                CST  
26300     America/New_York                      -5                EST  
26302      America/Chicago                      -6                CST  
26304     America/New_York                      -5                EST  

[2570 rows x 15 columns]

These duplicates are necessary

### Aggregating the skaters dataframe on team_id to pull insights from

In [27]:
# Aggregating the columns of interest in the skaters dataframe
df_dups_skaters.groupby('team_id').agg({'assists': ['sum'],'goals':['sum'],'shots':['sum'],'hits':['sum'],'blocked':['sum']})

assists goals shots    hits blocked
            sum   sum   sum     sum     sum
team_id                                    
1           662   404  4605  3305.0  2191.0
2           825   495  5252  5204.0  3032.0
3           771   456  4630  3731.0  2292.0
4           865   506  5182  3798.0  2380.0
5           854   504  5164  4496.0  2225.0
6           993   592  6134  4745.0  2518.0
7           690   414  4718  2882.0  1762.0
8           778   474  5476  4389.0  2224.0
9           725   432  4603  3761.0  2451.0
10          922   550  5451  2963.0  2185.0
12          874   518  5730  4023.0  2149.0
13          856   498  5033  3250.0  2155.0
14         1108   647  5792  5110.0  2528.0
15          910   543  5087  4165.0  2415.0
16          845   496  5099  3077.0  2278.0
17          625   366  4317  3327.0  2008.0
18          772   470  5345  2898.0  2296.0
19          951   564  5865  4205.0  2480.0
20          894   533  5313  3030.0  2212.0
21          993   589  5877  3826.0  2798.0
22          775   463  4599  3849.0  2284.0
23          820   491  5037  3681.0  2682.0
24          612   378  4377  3495.0  2139.0
25          832   499  5841  4880.0  3056.0
26          624   376  4682  3367.0  2031.0
28          914   527  5413  3784.0  2344.0
29          798   490  5515  3832.0  2533.0
30          737   438  4766  2848.0  2244.0
52          872   504  5044  3644.0  2185.0
53          667   421  4974  3654.0  2424.0
54          887   552  6231  5295.0  2559.0
87           25    17    53     1.0     2.0
88           29    22    64     0.0     3.0
89           15    10    40     0.0     3.0
90           24    15    48     0.0     3.0

In [28]:
# setting a variable to it
df_skate = df_dups_skaters.groupby('team_id').agg({'assists': ['sum'],'goals':['sum'],'shots':['sum'],'hits':['sum'],'blocked':['sum']})

# 3) Merging the data

In [29]:
# Merging df_skater_stats and df_game
df_merged = df_skater_stats.merge(df_game, on = 'game_id')

In [30]:
# setting it so i can see all columns
pd.set_option('display.max_columns', None)

In [31]:
# Running it
df_merged.head()

game_id  player_id  team_id  timeOnIce  assists  goals  shots  hits  \
0  2016020045    8468513        4        955        1      0      0   2.0   
1  2016020045    8476906        4       1396        1      0      4   2.0   
2  2016020045    8474668        4        915        0      0      1   1.0   
3  2016020045    8473512        4       1367        3      0      0   0.0   
4  2016020045    8471762        4        676        0      0      3   2.0   

   powerPlayGoals  powerPlayAssists  penaltyMinutes  faceOffWins  \
0               0                 0               0            0   
1               0                 0               2            0   
2               0                 0               0            0   
3               0                 2               0           11   
4               0                 0               0            0   

   faceoffTaken  takeaways  giveaways  shortHandedGoals  shortHandedAssists  \
0             0        1.0        1.0                 0                   0   
1             0        1.0        2.0                 0                   0   
2             0        2.0        0.0                 0                   0   
3            27        0.0        0.0                 0                   0   
4             0        0.0        1.0                 0                   0   

   blocked  plusMinus  evenTimeOnIce  shortHandedTimeOnIce  \
0      1.0          1            858                    97   
1      2.0          0           1177                     0   
2      0.0         -1            805                     0   
3      0.0         -1           1083                    19   
4      0.0         -1            613                    63   

   powerPlayTimeOnIce    season type         date_time_GMT  away_team_id  \
0                   0  20162017    R  2016-10-19T00:30:00Z             4   
1                 219  20162017    R  2016-10-19T00:30:00Z             4   
2                 110  20162017    R  2016-10-19T00:30:00Z             4   
3                 265  20162017    R  2016-10-19T00:30:00Z             4   
4                   0  20162017    R  2016-10-19T00:30:00Z             4   

   home_team_id  away_goals  home_goals       outcome home_rink_side_start  \
0            16           4           7  home win REG                right   
1            16           4           7  home win REG                right   
2            16           4           7  home win REG                right   
3            16           4           7  home win REG                right   
4            16           4           7  home win REG                right   

           venue           venue_link venue_time_zone_id  \
0  United Center  /api/v1/venues/null    America/Chicago   
1  United Center  /api/v1/venues/null    America/Chicago   
2  United Center  /api/v1/venues/null    America/Chicago   
3  United Center  /api/v1/venues/null    America/Chicago   
4  United Center  /api/v1/venues/null    America/Chicago   

   venue_time_zone_offset venue_time_zone_tz  
0                      -5                CDT  
1                      -5                CDT  
2                      -5                CDT  
3                      -5                CDT  
4                      -5                CDT

In [32]:
# Merging df_merged and df_team
df_merged_wins = df_merged.merge(df_team_stats, on = 'team_id')

In [33]:
# Show it
df_merged_wins.head()

game_id  player_id  team_id  timeOnIce  assists  goals  shots  hits  \
0  2016020045    8468513        4        955        1      0      0   2.0   
1  2016020045    8476906        4       1396        1      0      4   2.0   
2  2016020045    8474668        4        915        0      0      1   1.0   
3  2016020045    8473512        4       1367        3      0      0   0.0   
4  2016020045    8471762        4        676        0      0      3   2.0   

   powerPlayGoals  powerPlayAssists  penaltyMinutes  faceOffWins  \
0               0                 0               0            0   
1               0                 0               2            0   
2               0                 0               0            0   
3               0                 2               0           11   
4               0                 0               0            0   

   faceoffTaken  takeaways  giveaways  shortHandedGoals  shortHandedAssists  \
0             0        1.0        1.0                 0                   0   
1             0        1.0        2.0                 0                   0   
2             0        2.0        0.0                 0                   0   
3            27        0.0        0.0                 0                   0   
4             0        0.0        1.0                 0                   0   

   blocked  plusMinus  evenTimeOnIce  shortHandedTimeOnIce  \
0      1.0          1            858                    97   
1      2.0          0           1177                     0   
2      0.0         -1            805                     0   
3      0.0         -1           1083                    19   
4      0.0         -1            613                    63   

   powerPlayTimeOnIce    season type         date_time_GMT  away_team_id  \
0                   0  20162017    R  2016-10-19T00:30:00Z             4   
1                 219  20162017    R  2016-10-19T00:30:00Z             4   
2                 110  20162017    R  2016-10-19T00:30:00Z             4   
3                 265  20162017    R  2016-10-19T00:30:00Z             4   
4                   0  20162017    R  2016-10-19T00:30:00Z             4   

   home_team_id  away_goals  home_goals       outcome home_rink_side_start  \
0            16           4           7  home win REG                right   
1            16           4           7  home win REG                right   
2            16           4           7  home win REG                right   
3            16           4           7  home win REG                right   
4            16           4           7  home win REG                right   

           venue           venue_link venue_time_zone_id  \
0  United Center  /api/v1/venues/null    America/Chicago   
1  United Center  /api/v1/venues/null    America/Chicago   
2  United Center  /api/v1/venues/null    America/Chicago   
3  United Center  /api/v1/venues/null    America/Chicago   
4  United Center  /api/v1/venues/null    America/Chicago   

   venue_time_zone_offset venue_time_zone_tz  number_of_wins  
0                      -5                CDT            1735  
1                      -5                CDT            1735  
2                      -5                CDT            1735  
3                      -5                CDT            1735  
4                      -5                CDT            1735

In [34]:
# Checking for missing values
df_merged_wins.isnull().sum()

game_id                   0
player_id                 0
team_id                   0
timeOnIce                 0
assists                   0
goals                     0
shots                     0
hits                      0
powerPlayGoals            0
powerPlayAssists          0
penaltyMinutes            0
faceOffWins               0
faceoffTaken              0
takeaways                 0
giveaways                 0
shortHandedGoals          0
shortHandedAssists        0
blocked                   0
plusMinus                 0
evenTimeOnIce             0
shortHandedTimeOnIce      0
powerPlayTimeOnIce        0
season                    0
type                      0
date_time_GMT             0
away_team_id              0
home_team_id              0
away_goals                0
home_goals                0
outcome                   0
home_rink_side_start      0
venue                     0
venue_link                0
venue_time_zone_id        0
venue_time_zone_offset    0
venue_time_zone_tz  

### Merging df_team_stats and df_skate

In [35]:
df_relationships = df_skate.merge(df_team_stats, on = 'team_id')

E:\Anaconda\lib\site-packages\pandas\core\frame.py:9190: FutureWarning: merging between different levels is deprecated and will be removed in a future version. (2 levels on the left,1 on the right)
  return merge(


In [36]:
# running it
df_relationships

team_id  (assists, sum)  (goals, sum)  (shots, sum)  (hits, sum)  \
0         1             662           404          4605       3305.0   
1         2             825           495          5252       5204.0   
2         3             771           456          4630       3731.0   
3         4             865           506          5182       3798.0   
4         5             854           504          5164       4496.0   
5         6             993           592          6134       4745.0   
6         7             690           414          4718       2882.0   
7         8             778           474          5476       4389.0   
8         9             725           432          4603       3761.0   
9        10             922           550          5451       2963.0   
10       12             874           518          5730       4023.0   
11       13             856           498          5033       3250.0   
12       14            1108           647          5792       5110.0   
13       15             910           543          5087       4165.0   
14       16             845           496          5099       3077.0   
15       17             625           366          4317       3327.0   
16       18             772           470          5345       2898.0   
17       19             951           564          5865       4205.0   
18       20             894           533          5313       3030.0   
19       21             993           589          5877       3826.0   
20       22             775           463          4599       3849.0   
21       23             820           491          5037       3681.0   
22       24             612           378          4377       3495.0   
23       25             832           499          5841       4880.0   
24       26             624           376          4682       3367.0   
25       28             914           527          5413       3784.0   
26       29             798           490          5515       3832.0   
27       30             737           438          4766       2848.0   
28       52             872           504          5044       3644.0   
29       53             667           421          4974       3654.0   
30       54             887           552          6231       5295.0   
31       87              25            17            53          1.0   
32       88              29            22            64          0.0   
33       89              15            10            40          0.0   
34       90              24            15            48          0.0   

    (blocked, sum)  number_of_wins  
0           2191.0            1691  
1           3032.0            1743  
2           2292.0            1768  
3           2380.0            1735  
4           2225.0            1786  
5           2518.0            1822  
6           1762.0            1669  
7           2224.0            1732  
8           2451.0            1709  
9           2185.0            1709  
10          2149.0            1705  
11          2155.0            1682  
12          2528.0            1804  
13          2415.0            1789  
14          2278.0            1771  
15          2008.0            1713  
16          2296.0            1762  
17          2480.0            1794  
18          2212.0            1709  
19          2798.0            1731  
20          2284.0            1686  
21          2682.0            1736  
22          2139.0            1739  
23          3056.0            1761  
24          2031.0            1744  
25          2344.0            1787  
26          2533.0            1721  
27          2244.0            1714  
28          2185.0             888  
29          2424.0             650  
30          2559.0             468  
31             2.0               6  
32             3.0               6  
33             3.0               4  
34             3.0               4

In [ ]:
# Exporting the finished dataframe
# Exporting df_merged
df_merged.to_csv(os.path.join(path, '02 Data','Prepared', 'data_merged.csv'))

In [ ]:
# Exporting the finished dataframe
# Exporting df_relationship
df_relationships.to_csv(os.path.join(path, '02 Data','Prepared', 'data_relationship.csv'))

In [ ]:
# Exporting the finished dataframe
# Exporting df_merged_won
df_merged_wins.to_csv(os.path.join(path, '02 Data','Prepared', 'data_merged_wins.csv'))